---
title: "Indirect effects in DC-TAP-seq screens"
author: "Andreas R. Gschwind"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: show
---

```{r setupDocument, include=FALSE}
# set output html chunk options
knitr::opts_chunk$set(warning = FALSE)

# suppress dplyrs summarize info
options(dplyr.summarize.inform = FALSE)
```

```{r attachPackages, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggExtra)
library(cowplot)
library(DT)
```

## Goal
We empirically estimated the indirect effect rates in CRISPR experiments by testing distal elements
(DE) for trans-acting effects on genes on other chromosomes. By subtracting the indirect effect
rates from the observed cis positive rates, we modeled the expected direct effects rate as a
function of distance to TSS. Using the estimated indirect (\( r_{\text{indirect}}\)) and direct
effects rates as function of distance (\( r_{\text{direct}}(d) \)), we calculated a probability of
each CRISPR DE-gene pair to be the result of a direct effect as a function of distance to TSS:

$$
p_{\text{direct}}(d) = \frac{r_{\text{direct}}(d)}{r_{\text{direct}}(d) + r_{\text{indirect}}}
$$

Refer to the "Computing the probability that DE-G pair is a direct versus indirect effect" Methods
section of the manuscript for more information on this approach. Here we analyze the direct and
indirect rates in the K562 and WTC11 DC-TAP-seq screens to reproduce the plots shown in **Figure 4**
and **Figure S3**.

***

## Used data

This analysis uses the output of the
[**CRISPR indirect effects**](https://github.com/EngreitzLab/CRISPR_indirect_effects)
analysis workflow as input. It assumes that files generated by running it are located at:
```{r}
# adjust path to input files as needed to reproduce results
indirect_effects_dir <- "../../CRISPR_indirect_effects/results"
```

First we load all requied input files from the indirect effects analysis:
```{r loadData}
# load cis and trans positive rates from indirect effects analyses
cis_rates <- read_tsv(file.path(indirect_effects_dir, "cis_positive_hit_rates.tsv"),
                      show_col_types = FALSE)
trans_rates <- read_tsv(file.path(indirect_effects_dir, "trans_positive_hit_rates.tsv"),
                        show_col_types = FALSE)

# extract data on DC-TAP-seq screens only
cis_rates   <- filter(cis_rates,   dataset %in% c("K562_DC_TAPseq", "WTC11_DC_TAPseq"))
trans_rates <- filter(trans_rates, dataset %in% c("K562_DC_TAPseq", "WTC11_DC_TAPseq"))

# load main DC-TAP-seq results table
final_results_file <- file.path(indirect_effects_dir,
                                "annotated_crispr_data/EPCrisprBenchmark_dc_tap_allPairs_element_classes_direct_effects.tsv.gz")
final_results <- read_tsv(final_results_file, show_col_types = FALSE)

# load direct effects models
direct_effect_models <- readRDS(file.path(indirect_effects_dir, "direct_effect_models/direct_effects_model.rds"))
```

***

## Expected indirect hits
Calculate the percentage and number of indirect effects in cis DE-gene analyses for effects with a
negative, positive or any direction on target gene expression in the K562 and WTC11 screens.

First, get the expected number and percentage of expected indirect effects in the cis DE-gene
analysis.
```{r expectedIndirectEffectRates}
# filter cis hit rates to within 1Mb only
cis_rates <- filter(cis_rates, mean_dist < 1e6)

# count the number of cis pairs and hits per dataset across all distances and convert to long format
effect_rates <- cis_rates %>% 
  group_by(dataset) %>% 
  summarize(total_pairs = sum(total_pairs),
            significant_pairs = sum(significant_pairs),
            negative_pairs = sum(negative_pairs),
            positive_pairs = sum(positive_pairs),
            .groups = "drop") %>% 
  pivot_longer(cols = -c(dataset, total_pairs), names_to = "effect_direction",
               values_to = "significant_cis_pairs") %>% 
  select(dataset, effect_direction, tested_cis_pairs = total_pairs, significant_cis_pairs) %>% 
  mutate(effect_direction = sub("_pairs", "", effect_direction)) %>% 
  mutate(effect_direction = if_else(effect_direction == "significant", true = "any",
                                    false = effect_direction))

# convert indirect (trans) hits to long format and add to cis rates
effect_rates <- trans_rates %>% 
  select(dataset, tested_indirect_pairs = total_pairs, any = significant_pairs,
         negative = negative_pairs, positive = positive_pairs) %>% 
  pivot_longer(cols = -c(dataset, tested_indirect_pairs), names_to = "effect_direction",
               values_to = "significant_indirect_pairs") %>% 
  left_join(effect_rates, ., by = c("dataset", "effect_direction"))

# convert indirect (trans) hit rates to long format and add to cis rates
effect_rates <- trans_rates %>% 
  select(dataset, any = positive_rate_significant, negative = positive_rate_negative,
         positive = positive_rate_positive) %>% 
  pivot_longer(cols = -dataset, names_to = "effect_direction",
               values_to = "indirect_effects_rate") %>% 
  left_join(effect_rates, ., by = c("dataset", "effect_direction"))

# compute number and percentage of expected indirect effects
effect_rates <- effect_rates %>% 
  mutate(expected_indirect_cis_pairs = round(tested_cis_pairs * indirect_effects_rate)) %>% 
  mutate(percent_indirect_cis_pairs = expected_indirect_cis_pairs / significant_cis_pairs)
```

Next, compute how many of the final, filtered DE-gene pairs can be expected to be from indirect
effects based on the estimated percentage of indirect effects among all significant cis
DE-gene pairs.
```{r expectedIndirectEffectPairs}
# extract 'valid' random screen pairs within 1Mb
random_pairs <- final_results %>%
  filter((Significant | power_at_effect_size_15 >= 0.8)) %>% 
  filter(Random_DistalElement_Gene == TRUE) %>% 
  filter(abs(distanceToTSS) <= 1e6)

# get filtered final number of significant pairs
final_significant_pairs <- random_pairs %>%
  group_by(Dataset) %>%
  summarize(any = sum(Significant == TRUE),
         negative = sum(Significant == TRUE & EffectSize <= 0),
         positive = sum(Significant == TRUE & EffectSize > 0),
         .groups = "drop") %>% 
  pivot_longer(cols = -Dataset, names_to = "effect_direction",
               values_to = "filtered_significant_cis_pairs") %>% 
  mutate(Dataset = paste0(Dataset, "seq"))
  
# add to table with calculated effect rates
effect_rates <- left_join(effect_rates, final_significant_pairs,
                          by = c("dataset" = "Dataset", "effect_direction"))

# compute expected number of filtered pairs to be due to indirect effects based on
effect_rates <- effect_rates %>% 
  mutate(expected_indirect_final_cis_pairs = round(filtered_significant_cis_pairs * percent_indirect_cis_pairs))
```

```{r printTable1, echo=FALSE}
datatable(
  effect_rates,
  extensions = c("FixedColumns", "Buttons"),
  options = list(
    pageLength = 20,
    dom = 'Bfrtip',
    scrollX = TRUE,
    fixedColumns = list(leftColumns = 3),
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print')),
  autoHideNavigation = TRUE) %>%
  formatStyle(c(2, 4, 7, 9), `border-right` = "solid 1px")
```

*** 

## Direct vs indirect effects as function of distance
Next we show how the direct, indirect (trans) effect rates and probability of direct effects behave
as functions of distance to TSS

#### Overall cis hit and indirect hit rates
Here we show the overall (direct + indirect) cis hit rate of observed significant DE-gene pairs for
50kb bins of distance to TSS. The distance to TSS independent indirect effects rates are shown as
lines.
```{r cisVsIndirectEffectRate, fig.height=4, fig.width=8}
# reformat cis hit rates for plot
cis_rates_plot <- cis_rates %>% 
  select(dataset, mean_dist, positive_rate_negative, positive_rate_positive) %>% 
  pivot_longer(cols = -c(dataset, mean_dist), names_to = "effect_direction",
               values_to = "positive_rate") %>% 
  mutate(cell_type = sub("_DC_TAPseq", "", dataset)) %>% 
  mutate(effect_direction = str_to_title(sub("positive_rate_", "", effect_direction)))

# reformat trans hit rates for plot
trans_rates_plot <- trans_rates %>% 
  select(dataset, positive_rate_negative, positive_rate_positive) %>% 
  pivot_longer(cols = -dataset, names_to = "effect_direction", values_to = "positive_rate") %>% 
  mutate(cell_type = sub("_DC_TAPseq", "", dataset)) %>% 
  mutate(effect_direction = str_to_title(sub("positive_rate_", "", effect_direction)))  

# plot cis and trans hit rates as function of distance to TSS
ggplot(cis_rates_plot,
       aes(x = mean_dist / 1e6, y = positive_rate, color = cell_type, shape = effect_direction)) +
  facet_wrap(~cell_type, scales = "free") +
  geom_point() +
  geom_hline(data = trans_rates_plot, aes(yintercept = positive_rate, linetype = effect_direction,
                                          color = cell_type)) +
  labs(title = "Positive hit rates cis and trans analyses", x = "Distance to TSS (Mb)",
       y = "Positive hite rate", color = "Cell type", shape = "Effect direction",
       linetype = "Effect direction") +
  scale_color_manual(values = c(K562 = "#b00404", WTC11 = "#0202cc")) +  
  scale_shape_manual(values = c(Negative = 19, Positive = 1)) +
  scale_y_continuous(limits = c(NA, max(cis_rates_plot$positive_rate)), labels = scales::percent) +
  theme_classic() +
  theme(strip.background = element_blank())
```

```{r savePlot1, echo=FALSE}
# save plot to pdf
ggsave(file = "./plots//cis_and_trans_effect_rates.pdf", height = 4, width = 8)
```

#### Probability of direct effect
The overall cis and indirect hit rates are used to compute the probability of direct effects as
described above. Note that at this stage we included results from additional CRISPR enhancer
perturbation screens to compute the expected direct effects rate as function of TSS (see Methods).

This plot shows the probability of direct effects as function of distance to TSS for the K562 and
WTC11 DC-TAP-seq screens.
```{r pDirectVsDistance, fig.height=4, fig.width=6}
# add effects direction label for each CRISPR E-G pair
random_pairs <- random_pairs %>% 
  mutate(effect_direction = if_else(EffectSize <= 0, true = "Negative", false = "Positive"))

# select positive and negative indirect rates and convert to long format for plot
random_pairs_plot <- random_pairs %>% 
  select(element_gene_pair_identifier_hg38, cell_type, dist_to_tss, Significant,
         direct_vs_indirect_negative, direct_vs_indirect_positive, effect_direction) %>% 
  pivot_longer(cols = c(direct_vs_indirect_negative, direct_vs_indirect_positive),
               names_to = "rate_type", values_to = "direct_vs_indirect") %>% 
  mutate(rate_type = str_to_title(sub("direct_vs_indirect_", "", rate_type))) %>% 
  mutate(label = paste0(cell_type, " (", rate_type, " effects)"))

# plot probability of direct effects as function of distance to TSS
ggplot(random_pairs_plot, aes(x = dist_to_tss / 1e3, y = direct_vs_indirect, color = cell_type,
                              linetype = rate_type, shape = effect_direction)) +
  geom_line() +
  geom_point(data = filter(random_pairs_plot, Significant == TRUE, effect_direction == "Negative",
                           rate_type == "Negative")) +
  geom_point(data = filter(random_pairs_plot, Significant == TRUE, effect_direction == "Positive",
                           rate_type == "Positive")) +  
  labs(title = "Probability direct vs. indirect effect", x = "Distance to TSS (kb)",
       y = "Probability direct vs. indirect effect", color = "Cell type",
       linetype = "Effect direction", shape = "Effect direction") + 
  scale_linetype_manual(values = c(Negative = 1, Positive = 3)) +
  scale_color_manual(values = c(K562 = "#b00404", WTC11 = "#0202cc")) +
  scale_shape_manual(values = c(Negative = 19, Positive = 1)) +
  theme_classic()
```

```{r savePlot2, echo=FALSE}
# save plot to pdf
ggsave(file = "./plots//probability_direct_vs_indirect.pdf", height = 4, width = 6)
```

***

## Properties of likely direct and indirect DE-gene pairs
Next we use the probability of direct effects to analyze properties of DE-gene pairs that are
likely to be the result of direct or indirect effects. In these analyses we only look at the
probability of direct effects for negative effects on gene expression (downregulation), because
the K562 and WTC11 experiments were designed to target enhancers, which are expected to have a
negative effect on target gene expression when perturbed.

#### Distance to TSS
Using the model for direct effects as function of distance to TSS and the estimated indirect
effects, we can calculate at which distance to TSS CRISPR hits have >90% and <20% probability to be
direct effects.
```{r definePDirectDistFunction}
# function to predict the direct hit rate for a given distance using the fitted models for direct
# effects as function of distance to TSS
predict_direct_hit_rate <- function(models, new_data, type) {
  
  # get model for specified type
  fit <- models[[type]]
  
  # predict direct hit rate for provided distances and exponentiate to get back to original scale
  predicted_log_direct_rate <- predict(fit, newdata = new_data)
  predicted_direct_rate <- exp(predicted_log_direct_rate)
  
  # cap predicted values at 1, because the rate can't be higher than that
  predicted_direct_rate <- pmin(predicted_direct_rate, 1)
  
  return(predicted_direct_rate)
  
}
```

```{r calculatePDirectDist}
# create table with 1bp to 1Mb distance to TSS
direct_effects <- bind_rows(K562_DC_TAPseq = tibble(dist_to_tss = 1:1e6),
                            WTC11_DC_TAPseq = tibble(dist_to_tss = 1:1e6),
                            .id = "dataset")

# predict direct effects rate across all distances
direct_effects <- direct_effects %>% 
  mutate(direct_hit_rate = predict_direct_hit_rate(direct_effect_models, new_data = direct_effects,
                                                   type = "direct_rate_negative"))

# add indirect effects rate for each dataset
direct_effects <- trans_rates %>% 
  select(dataset, indirect_hit_rate = positive_rate_negative) %>% 
  left_join(direct_effects, ., by = "dataset")

# calculate probability direct vs. indirect hit
direct_effects <- direct_effects %>% 
  mutate(direct_vs_indirect =  direct_hit_rate / (direct_hit_rate + indirect_hit_rate))
```

```{r calculateDistAtPDirect}
# get distance to tss and probability direct vs. indirect effect and convert to wide format
prob_direct_vs_indirect <- direct_effects %>% 
  select(dataset, dist_to_tss, direct_vs_indirect) %>% 
  pivot_wider(names_from = dataset, values_from = direct_vs_indirect)

# function to calculate distance at which E-G pairs are above or below a certain probability
get_dist_direct_prob <- function(x, col, prob, sign = c(">", "<")) {
  if (sign == ">") {
    max(x[x[[col]] > prob, "dist_to_tss"])
  } else if(sign == "<") {
    min(x[x[[col]] < prob, "dist_to_tss"])
  } else {
    stop("Invalid 'sign' argument", call. = FALSE)
  }
}

# calculate distance of E-G pairs at which probability of being a direct effect is >0.9 and <0.2 
dist_direct_probs <- tibble(
  `K562.>90` = get_dist_direct_prob(prob_direct_vs_indirect, "K562_DC_TAPseq", 0.9, ">"),
  `WTC11.>90` = get_dist_direct_prob(prob_direct_vs_indirect, "WTC11_DC_TAPseq", 0.9, ">"),
  `K562.<20` = get_dist_direct_prob(prob_direct_vs_indirect, "K562_DC_TAPseq", 0.2, "<"),
  `WTC11.<20` = get_dist_direct_prob(prob_direct_vs_indirect, "WTC11_DC_TAPseq", 0.2, "<") 
)

# convert to long format and reformat for nice printing
dist_direct_probs <- dist_direct_probs %>% 
  pivot_longer(cols = everything(), names_to = "stat", values_to = "Distane to TSS (bp)") %>%
  mutate(`Effect direction` = "Negative", .before = 1) %>% 
  separate(stat, into = c("Cell type", "Probability"), sep = "\\.")
```

```{r printTable2, echo=FALSE}
datatable(dist_direct_probs)
```

##### Enrichment for positive effects among distal pairs
Using the 90% probability of direct effects cutoff, we can see that pairs below that threshold are
enriched for positive effects on expression. Enhancer-gene interactions are expected to have
negative effects on target gene expression when perturbed. Therefore, the higher percentage of
positive effects on expression among DE-gene pairs with lower probability of being direct effects,
supports that these have a higher proportion of effects from indirect, e.g. trans-acting mechanisms.
```{r countSigPDirect}
# annotate results of direct vs indirect effect probability is >0.9 or <0.2
random_pairs <- random_pairs %>% 
  mutate(p90 = if_else(direct_vs_indirect_negative > 0.9, true = "Prob. direct > 0.9",
                       false = "Prob. direct <= 0.9"))

# count number of significant pairs with positive and negative effects for pairs with direct vs
# indirect effect probability greater or lower than 0.9
pairs_p90 <- random_pairs %>% 
  group_by(cell_type, p90) %>% 
  summarize(significant = sum(Significant == TRUE),
            negative = sum(Significant == TRUE & EffectSize <= 0),
            positive = sum(Significant == TRUE & EffectSize > 0),
            prop_negative = negative / significant,
            prop_positive = positive / significant)
```

Plot the proportions of significant positive and negative effects for K562 and WTC11 experiments
separately.
```{r plotEffectDirectionPDirect1, fig.height=4, fig.width=5}
# transform to long format for plotting
pairs_p90_plot <- pairs_p90 %>% 
  mutate(plot_label = paste0(p90, " (n=", significant, ")")) %>% 
  select(cell_type, plot_label, prop_negative, prop_positive) %>% 
  pivot_longer(cols = c(prop_negative, prop_positive), names_to = "Effect direction",
               values_to = "% significant pairs") %>% 
  mutate(`Effect direction` = str_to_title(sub("prop_", "", `Effect direction`)))

# plot the proportion of negative vs positives based on probability to be a direct effect
ggplot(pairs_p90_plot, aes(x = fct_rev(plot_label), y = `% significant pairs`, fill = `Effect direction`)) +
  facet_wrap(~ cell_type, scales = "free", drop = TRUE) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0.5, size = 0.25) +
  labs(title = "Negative vs. postive effects") +
  scale_fill_manual(values = c(Negative = "steelblue", Positive = "firebrick3")) +
  scale_y_continuous(labels = scales::percent) +
  theme_classic() +
  theme(axis.title.x = element_blank(), strip.background = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```

Plot proportions for both experiments combined.
```{r plotEffectDirectionPDirect2, fig.height=4, fig.width=3.5}
# count number of significant pairs with positive and negative effects for pairs with direct vs
# indirect effect probability greater or lower than 0.9
pairs_p90_combined <- random_pairs %>% 
  group_by(p90) %>% 
  summarize(significant = sum(Significant == TRUE),
            negative = sum(Significant == TRUE & EffectSize <= 0),
            positive = sum(Significant == TRUE & EffectSize > 0),
            prop_negative = negative / significant,
            prop_positive = positive / significant)

# transform to long format for plotting
pairs_p90_combined_plot <- pairs_p90_combined %>% 
  mutate(plot_label = paste0(p90, " (n=", significant, ")")) %>% 
  select(plot_label, prop_negative, prop_positive) %>% 
  pivot_longer(cols = c(prop_negative, prop_positive), names_to = "Effect direction",
               values_to = "% significant pairs") %>% 
  mutate(`Effect direction` = str_to_title(sub("prop_", "", `Effect direction`)))

# plot the proportion of negative vs positives based on probability to be a direct effect
ggplot(pairs_p90_combined_plot,
       aes(x = fct_rev(plot_label), y = `% significant pairs`, fill = `Effect direction`)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = 0.5, size = 0.25) +
  labs(title = "Negative vs. postive effects") +
  scale_fill_manual(values = c(Negative = "steelblue", Positive = "firebrick3")) +
  scale_y_continuous(labels = scales::percent) +
  theme_classic() +
  theme(axis.title.x = element_blank(), strip.background = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r, savePlot3, echo=FALSE}
# save plot to pdf
ggsave(file = "./plots//negative_vs_positive_effects_combined.pdf", height = 4, width = 3.5)
```

Alternatively, we can also look at this by plotting the observed CRISPR effect size on target gene
expression as a function of probability to be direct vs. indirect effect.
```{r plotEffectSizePDirect, fig.height=4, fig.width=6, eval = FALSE}
# get all significant hits
random_pairs_hits <- filter(random_pairs, Significant == TRUE)

# plot CRISPR effect size of likely direct vs. likely indirect effects
ggplot(random_pairs_hits, aes(x = cell_type, y = log_2_FC_effect_size, color = fct_rev(p90),
                                 group = interaction(cell_type, fct_rev(p90)))) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.5) +
  geom_point(alpha = 0.5, position = position_jitterdodge(jitter.width = 0.3)) +
  geom_boxplot(color = "black", fill = NA, outlier.shape = NA) +
  labs(title = "CRISPR effect size likely direct vs. indirect enhancer effect",
       y = "CRISPR effect size\n(log2 fold change)", color  = "Probability\ndirect vs indirect") +
  scale_color_manual(values = c(`Prob. direct > 0.9` = "firebrick3", `Prob. direct <= 0.9` = "steelblue")) +
  theme_classic() +
  theme(axis.title.x = element_blank())
```

```{r savePlot4, echo=FALSE}
# save plot to pdf
ggsave(file = "./plots//probability_direct_vs_effect_size.pdf", height = 4, width = 6)
```

***

## Session information
This document was produced using following packages:
```{r sessionInfo}
sessionInfo()
```
